CMakeLists.txt

```diff
@@ -1,98 +1,21 @@
-# $Id: $
+cmake_minimum_required(VERSION 3.12.0)
```

Comment:
As per discussion, 3.12 is required for PA_BUILD_TESTS to succeed in case of shared library build scenario but in this case tests do not test actual shared library binary after it was assembled by the linker which could introduce some incorrect binary code due to optimization and the internal linker bug. In this case tests will not reflect actual operation of the shared library binary as it is not being tested and will provide misrepresentation of the result (actual shared library binary has to participate in the tests).

Therefore requiring CMake 3.12 is excessive and need to be lowered to 3.0 or even 2.8 if other used CMake APIs allow that.

Linking object files directly is the same as linking the static library in this case and therefore the PA_BUILD_TESTS could be limited to a static build scenario (`BUILD_SHARED_LIBS = OFF`) and later extended by another PR to a shared library once the functions needed for tests are exported from PortAudio library and become available for linking from a shared library.

Reply:
Since @Be-ing wrote the new `CMakeLists.txt` and knows the internals best, he should answer here. I cannot comment on the internals. If @Be-ing defends his current solution I would support him regarding the CMake version since object libraries are a widely supported standard feature in CMake in non-out-dated versions.

The context of this discussion is a little misleading. This discussion is not about:

https://github.com/PortAudio/portaudio/blob/c4e1544ad335e9968d6aa63aaa856f39a418fba5/CMakeLists.txt#L1

but rather this:

https://github.com/PortAudio/portaudio/blob/c4e1544ad335e9968d6aa63aaa856f39a418fba5/CMakeLists.txt#L12-L17

Reply:
@h3ndrk you are right, this is what I am trying to say - there is no need for this workaround (linking against object files) because tests are testing implementation statically linked to them as object files, not a static or shared binary.  Tests should link either static or shared binary that will guarantee that binary under the test has passed them.

Once it is understood you will realize that Object Library feature of CMake 3.12+ will not be used and as a result CMake min version requirement can be reverted back to 3.0.

Reply:
Don't get me wrong: I'm not saying that there is no need for this workaround. If @Be-ing wants to use private functions in the tests he needs to include them into the binary. IIrc one fully legitimate option to do that is with object libraries. Since we live and develop with today's software (non-outdated CMake versions) we should use all the tools that are available if they fit the task well. Object libraries may be one of them. If this requires a newer version of CMake then download it simply. The developers should not think about how it could be worked around with other ways if object libraries are the way to go. There is absolutely no need to reinvent the wheel and waste developer's time.

And again: For setting it back to 3.0 or whatever, you need to test whether every little detail works in the same way in which @Be-ing once developed it with. Good luck with that, this can be very hard. It is discouraged by very smart people.